### ISSUE TITLE
Emote rendering and spacing behavior differs from native Kick functionality

### ISSUE THREAD

**Issue by pwnflakes** [Xzensi/NipahTV#187]
Without Nipah the emote forms/displays on the last char of the emote, not requiring a spacebar to form it, meaning the space works/creates a space instead of just completing the emote. Well actually it adds an extra space because Kick auto-adds a space after the emote registers, so when hitting space there's 2 in the box but Kick strips extra spaces.

I'm not sure what the most 'expected' way this should work, because funnily both Kick and Twitch do it differently, and Twitch similar to how Nipah has.

On Kick when you type a native emote name, it converts to an image in the chatbox on the last character entry & adds a space after the emote without even hitting space. This has its issues.

On Twitch when you type a native emote name, like Nipah it also doesn't convert into an image until you hit space, however it shows a tooltip above the text with the emote image, and when you do hit space it converts the text into the emote image and adds a space after it.

Both their implementations are decent, at least creating the next space in the same action after a valid emote is typed. I guess part of the issue is deciding what do you do in the scenario that an emote called 'EZ' exists and the user wants to type "EZRA", would it render an emote+RA?  or send the text
Well on Kick atm in this scenario it would render the emote which can be bad, however no official kick emotes are likely to be accidentally used where they were meant to be used as part of a word, but of course the extension handles more than official emotes. Twitch's method seems the best of both worlds as it doesn't convert until you hit space, but you know what the emote is before hitting space still cause of the tooltip. This of course allows you to type a word/something else instead of the emote, but doesnt require two spaces to give the feedback of a space being present. 

Also interestingly even though Nipah doesn't display a space, one is added even if you don't do that 2nd space. This also means, because there's nothing to 'remove/backspace' you aren't able to send a native emote with no space between it and the text, and this is possible extensionless, so maybe adding that space after rendering emote is a good idea so that when you backspace (if you wanted to) it could put the emote right next to the text 

First below is native, the bottom message is a result of typing the emote and hitting backspace after to remove the space they added as soon as the last char was typed. The first is just what kick does if typing literally "beeBobbletest" with no space, it ofc adds one.

Native:
![image](https://github.com/user-attachments/assets/c706899b-ad80-4e97-b4bc-27d86809d87c)

In Nipah it'll always show a space even though there's nothing to delete to bring the emote closer to the next text
Nipah:
![image](https://github.com/user-attachments/assets/e5580886-2574-4b5f-9bbd-1a6ea6b73063)

One message sent with Nipah, and one Native, with no spaces between them, viewed on extensionless:
Top line is what Nipah sent (even though no spaces were present)
Bottom line is what native Kick sent when you manually remove the spaces between the rendered emotes, showing the emotes closer together.
![image](https://github.com/user-attachments/assets/d4c1236e-9dde-40a5-b798-4e92634ee93c)

From the Nipah view, looking at that same scenario, both messages seem pretty equal here instead. So in a way the issue isn't just with entry, but also with Nipah actually not kind of displaying what's "truly" sent when no spaces are used.
![image](https://github.com/user-attachments/assets/2e75e4a4-5541-4c29-88cd-e1263f26e403)

I probably didn't need to get this detailed on this but hey, and they also arent huge issues

**Comment by Xzensi** [comment]
You're right in observing this behaviour and it has been a conscious design decision due the following reasoning.

First of all it's important to consider that Kick does not respect double spaces. Meaning it's not possible at any point to send 2 consecutive spaces. Secondly, we have to consider that depending on the emote format, for some emote providers we can handle the emotes without spaces, and others at least one dividing space is required to be able to denote a single isolated emote in a text string.

For Kick, the format is e.g. "[emote:1234:EZ]" and for 7TV (and BTTV) it is "EZ". Because of this 7TV/BTTV emote format it is impossible to single out emotes without spaces e.g. "EZEZEZ", therefore these emotes always require a space "EZ EZ EZ". As result, it is technically impossible on Kick to have extra spaces between 7TV/BTTV emotes. While we could technically add spaces between Kick emotes and still be able to parse it correctly, it'd mean that we'd be able to have Kick emotes right next to each other with no spacing but just not for 7TV/BTTV emotes.

I didn't like how this resulted in messy and misaligned emotes in chat messages where some emotes would have no spacing and some did, and on top of that it was a pain to implement correctly due to the requirements of some emotes requiring spaces and some not and handling it correctly for every case (although it does work correctly and I could revert this decision).

That's why fairly early on I made the design decision to just get rid of all spaces between emotes completely so there's no need for spaces at all. That being said, I am planning to completely trash the custom rich content editor I wrote and replace it with an existing library to fix various accessibility issues, and maybe I'll revisit and re-evaluate this decision then.

It's just a bit sad that there's many third party emotes that look great with no spacing between the emotes, but it'd be impossible to support this as all these emotes would never have any spacing at all. Unless, I completely broke backwards compatibility with the 7TV/BTTV extensions and other third party extensions by changing the emote format to "[EZ]" or the like, but we don't have the critical mass in adoption to make this worth considering.

**Comment by Xzensi** [comment]
Closing due to inactivity
